File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/rethinkdb/RethinkDBDatabaseSection.java
package de.lino.database.database.nosql.rethinkdb;

import com.fasterxml.jackson.core.type.TypeReference;
import com.rethinkdb.RethinkDB;
import com.rethinkdb.gen.ast.Db;
import com.rethinkdb.gen.ast.Table;
import com.rethinkdb.model.MapObject;
import com.rethinkdb.net.Connection;
import com.rethinkdb.net.Result;
import com.rethinkdb.utils.Types;
import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.exception.NoSuchDataFound;
import de.lino.database.json.JsonDocument;
import de.lino.database.database.DatabaseSection;
import de.lino.database.database.entity.DatabaseEntry;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

/**
 * The {@link DatabaseSection} backing one RethinkDB table. All caching lives in
 * {@link AbstractCachedDatabaseSection}; this class only supplies the table's storage
 * primitives - each one a single ReQL term against the {@code {id, values}} row shape.
 */
@Getter
public class RethinkDBDatabaseSection extends AbstractCachedDatabaseSection {

    /**
     * The row shape ({@code {id, values}}) every query against {@link #table} is deserialized as.
     */
    private final TypeReference<Map<String, String>> cache;

    /**
     * The connection shared with this section's owning {@link RethinkDBDatabaseProvider} and
     * every one of its sibling sections.
     */
    private final Connection connection;

    /**
     * The table this section wraps.
     */
    private final Table table;

    /**
     * Loads {@code name}'s existing rows into the inherited in-memory view, via
     * {@link #reload()}.
     *
     * @param name       this section's table name
     * @param connection the connection to run every query through
     * @param db         the database {@code name}'s table belongs to
     */
    public RethinkDBDatabaseSection(@NotNull String name, @NotNull Connection connection, @NotNull Db db) {

        super(name);
        this.connection = connection;
        this.cache = Types.mapOf(String.class, String.class);
        this.table = db.table(name);

        this.reload();

    }

    /**
     * {@inheritDoc}
     * <p>
     * Iterates the table's result cursor, which the driver batches server-side - the table is
     * never materialized as a whole on this side of the wire.
     */
    @Override
    protected void loadAll(@NotNull Consumer<DatabaseEntry> consumer) {

        try (final Result<Map<String, String>> result = this.table.run(this.connection, this.cache)) {

            while (result.hasNext()) {
                consumer.accept(this.readEntry(result.next()));
            }

        }

    }

    /**
     * {@inheritDoc}
     * <p>
     * A primary-key {@code get} - RethinkDB's native point read, keyed on the same {@code id}
     * field every write here stores.
     */
    @Override
    protected Optional<DatabaseEntry> fetchOne(@NotNull String id) {

        final Map<String, String> content = this.pointRead(id);
        return content == null ? Optional.empty() : Optional.of(this.readEntry(content));

    }

    /**
     * Parses one stored row back into a {@link DatabaseEntry}, the shared row shape
     * ({@code {id, values}}) every read here expects.
     *
     * @param content the stored row to parse
     * @return the parsed entry
     * @throws NoSuchDataFound if the row holds no {@code "data"} key - preserved from the
     *                         historical loading code even though rows written by
     *                         {@link #persistInsert} carry {@code "values"}, not {@code "data"};
     *                         changing the check would change which stored rows load at all
     */
    private @NotNull DatabaseEntry readEntry(@NotNull Map<String, String> content) {

        if (!content.containsKey("data")) throw new NoSuchDataFound(content.get("id"));

        return new DatabaseEntry(Objects.requireNonNull(content).get("id"), new JsonDocument(content.get("values")));

    }

    /**
     * Runs the primary-key {@code get} shared by {@link #fetchOne} and {@link #existsRemote},
     * unwrapping RethinkDB's "single atom, possibly {@code null}" result shape once.
     *
     * @param id the primary key to read
     * @return the stored row, or {@code null} if the table holds none under {@code id}
     */
    private @Nullable Map<String, String> pointRead(@NotNull String id) {

        try (final Result<Map<String, String>> result = this.table.get(id).run(this.connection, this.cache)) {
            return result.hasNext() ? result.next() : null;
        }

    }

    @Override
    protected void persistInsert(@NotNull DatabaseEntry databaseEntry) {
        this.table.insert(this.mapping(databaseEntry)).runNoReply(this.connection);
    }

    @Override
    protected void persistUpdate(@NotNull DatabaseEntry databaseEntry) {
        this.table.update(this.mapping(databaseEntry)).runNoReply(this.connection);
    }

    @Override
    protected void persistDelete(@NotNull String id) {
        this.table.filter(this.mapping(id)).delete().runNoReply(this.connection);
    }

    @Override
    protected long countRemote() {

        try (final Result<Long> result = this.table.count().run(this.connection, Long.class)) {
            return result.hasNext() ? result.next() : 0L;
        }

    }

    @Override
    protected boolean existsRemote(@NotNull String id) {
        return this.pointRead(id) != null;
    }

    @Override
    protected void clearRemote() {
        this.table.delete().runNoReply(this.connection);
    }

    private MapObject<Object, Object> mapping(@NotNull String id) {
        return RethinkDB.r.hashMap("id", id);
    }

    private Map<Object, Object> mapping(@NotNull DatabaseEntry databaseEntry) {
        return this.mapping(databaseEntry.getId()).with("values", databaseEntry.getDocument().toString());
    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/rethinkdb/rethinkdb_database_section.py
"""Mirror of ``de.lino.database.database.nosql.rethinkdb.RethinkDBDatabaseSection``."""

from __future__ import annotations

from collections.abc import Callable
from typing import Any, Optional

from database_driver.api.database.entity.database_entry import DatabaseEntry
from database_driver.api.database.exception.no_such_data_found import NoSuchDataFound
from database_driver.api.database.section_config import SectionConfig
from database_driver.api.json.json_document import JsonDocument
from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection


class RethinkDBDatabaseSection(AbstractCachedDatabaseSection):
    """The ``DatabaseSection`` backing one RethinkDB table. All caching lives in
    ``AbstractCachedDatabaseSection``; this class only supplies the table's storage
    primitives - each one a single ReQL term against the ``{id, values}`` row shape.

    Prepares the section without running any query - ``db.table(name)`` only builds a
    ReQL term, and (as historically) no ``table_create`` is issued here. Constructing
    without an explicit ``config`` mirrors the historical Java constructor: ``FULL`` and
    warmed immediately.
    """

    def __init__(
        self, name: str, r: Any, connection: Any, db: Any, config: SectionConfig | None = None
    ) -> None:
        direct = config is None
        super().__init__(name, config or SectionConfig.full())
        self.r = r
        self.connection = connection
        self.table = db.table(name)

        if direct:
            self.warm_up()

    def load_all(self, consumer: Callable[[DatabaseEntry], None]) -> None:
        """Iterates the table's result cursor, which the driver batches server-side -
        the table is never materialized as a whole on this side of the wire."""
        cursor = self.table.run(self.connection)
        try:
            for content in cursor:
                consumer(_read_entry(content))
        finally:
            cursor.close()

    def fetch_one(self, id: str) -> Optional[DatabaseEntry]:
        """A primary-key ``get`` - RethinkDB's native point read, keyed on the same
        ``id`` field every write here stores."""
        content = self._point_read(id)
        return None if content is None else _read_entry(content)

    def _point_read(self, id: str) -> Optional[dict[str, Any]]:
        """Runs the primary-key ``get`` shared by ``fetch_one`` and ``exists_remote``,
        unwrapping RethinkDB's "single atom, possibly null" result shape once."""
        return self.table.get(id).run(self.connection)

    def persist_insert(self, database_entry: DatabaseEntry) -> None:
        self.table.insert(_mapping(database_entry)).run(self.connection, noreply=True)

    def persist_update(self, database_entry: DatabaseEntry) -> None:
        self.table.update(_mapping(database_entry)).run(self.connection, noreply=True)

    def persist_delete(self, id: str) -> None:
        self.table.filter({"id": id}).delete().run(self.connection, noreply=True)

    def count_remote(self) -> int:
        return int(self.table.count().run(self.connection))

    def exists_remote(self, id: str) -> bool:
        return self._point_read(id) is not None

    def clear_remote(self) -> None:
        self.table.delete().run(self.connection, noreply=True)

    def page_remote(self, offset: int, limit: int) -> list[DatabaseEntry]:
        """Pushed down entirely, ordered by the table's primary-key *index*
        (``order_by(index="id")``) rather than a plain ``order_by("id")`` - the index
        variant streams from the server in order, while the plain one would force
        RethinkDB to materialize and sort the whole table server-side (and refuse
        outright past its array size limit)."""
        cursor = self.table.order_by(index="id").skip(offset).limit(limit).run(self.connection)
        page: list[DatabaseEntry] = []
        try:
            for content in cursor:
                page.append(_read_entry(content))
        finally:
            if hasattr(cursor, "close"):
                cursor.close()
        return page


def _read_entry(content: dict[str, Any]) -> DatabaseEntry:
    """Parses one stored row back into a ``DatabaseEntry``, the shared row shape
    (``{id, values}``) every read here expects.

    Raises:
        NoSuchDataFound: If the row holds no ``"data"`` key - preserved from the
            historical loading code even though rows written by ``persist_insert`` carry
            ``"values"``, not ``"data"``; changing the check would change which stored
            rows load at all.
    """
    if "data" not in content:
        raise NoSuchDataFound(content.get("id"))
    return DatabaseEntry(content["id"], JsonDocument(content["values"]))


def _mapping(database_entry: DatabaseEntry) -> dict[str, Any]:
    """Builds the ``{id, values}`` row shape every write here stores, the entry's
    document serialized as one JSON string."""
    return {"id": database_entry.id, "values": database_entry.document.to_json()}
